### ISSUE TITLE
Shortcut for generating AST fragments

### ISSUE THREAD

**Issue by overlookmotel** [oxc-project/oxc#4112]
This is the code we have to write in transformer to generate `let Foo;`:

```rs
let kind = VariableDeclarationKind::Let;
let declarations = {
    let ident = BindingIdentifier::new(SPAN, name.clone());
    let pattern_kind = self.ctx.ast.binding_pattern_identifier(ident);
    let binding = self.ctx.ast.binding_pattern(pattern_kind, None, false);
    let decl = self.ctx.ast.variable_declarator(SPAN, kind, binding, None, false);
    self.ctx.ast.new_vec_single(decl)
};
Declaration::VariableDeclaration(self.ctx.ast.variable_declaration(
    SPAN,
    kind,
    declarations,
    Modifiers::empty(),
))
```

Seems a bit bananas! Very long to write, and non-trivial to understand at first glance what the code is doing.

Surely we can find some way to reduce this.

Like: `ast!{ let #name; }`

The obvious problem is compile time impact of macros. Can we find some way to precompile macro expansion so we can enjoy a simpler code style without paying the compile-time price?

**Comment by overlookmotel** [comment]
Me, @Boshen, @rzvxa and @Dunqing discussed this today on a video call.

### How SWC does this

We noted that SWC does something similar to the `ast!` macro mentioned above. Their version is:

```rs
let stmt = quote!{
    "let $name = true;" as Stmt,
    name = BindingIdentifier { span: SPAN, name: "Foo" }
};
```

We decided that we'd prefer to have the template insertions outside of the macro like this:

```rs
let name = BindingIdentifier { span: SPAN, name: "Foo" };
let stmt = ast_stmt!{ let #name = 123; };
```

### Our plan

We agreed that:

* Ideal for the user is the `ast!` macro approach (compared to the other options of AST snippets, or adding loads of complicated methods to AST builder).
* We should try to do that.
* Parsing the JS code fragment inside the macro will happen at compile time.
* Macro will expand to `Statement::VariableDeclaration( /* ...etc... */ )`.
* Initial implementation will hurt compile times due to the macros.
* In a 2nd phase, we will improve compile time performance using some form of caching (as per rzvxa's comment above).
* rzvxa is going to take the lead on this.

I believe that's an accurate representation of our discussion. If not, please let me know!

**Comment by Boshen** [comment]
nit: `#name` is a private identifier. We'll let the implementor to decide what marker to use.

**Comment by overlookmotel** [comment]
> Note1: `#name` is a private identifier.

Good point. Is there another "special" character we can use which is unambiguous? `%name`? `@@name`?

**Comment by rzvxa** [comment]
Maybe we can use a ruby-like interpolation:
```rust
ast!(let #{name} = 123;)
```
It won't conflict with any valid js code, And it can't be mistaken with ``` `${name}` ```, We can also use the same syntax with parentheses for expanding iterators.

```rust
ast!(let arr = [ #(#{elements}),* ])
```

Both `#(ident)` and `#{ident}` are syntax errors as of now so maybe we can leverage them to our advantage But [this proposal](https://github.com/tc39/proposal-record-tuple) can cause problems in the future if it gets in.
https://tc39.es/proposal-record-tuple/#sec-ecmascript-language-lexical-grammar

With that said IMHO, The odds of it getting in with this syntax is slim. TC39 nowadays is trying to be closer to the typescript and TS folks would prefer a C# like syntax with `Tuple/Record`; So they would prefer this:
```js
let a = Tuple(1, 2, 3);
```

Over this:

```js
let a = #[1, 2, 3];
```

But it's just a wild guess.

**Comment by overlookmotel** [comment]
I like the Ruby-style interpolation syntax. It allows more flexibility than `quote!`'s interpolation which can only accept an identifier `$foo`, whereas with this style we can do `#{foo.bar}`, `#{self.foo.bar(qux)}` etc.

Shall we go with `#(...)` just to avoid potential clash with Tuple syntax?

Sequence expansion could use `#(..#(elements)),*` or `##(#(elements)),*` (or something else, many possibilities).

I'm not dead set on this - `#{...}` works for now.

**Comment by rzvxa** [comment]
The reason I chose this is because of its familiarity, It looks like `${expr}` to the people, they have to use `#` over `$`.
But going with parenthesis might be worth it to avoid any possible clash. It is such a minor change that we can discuss it at the last minute and still be fine.

**Comment by overlookmotel** [comment]
Actually, `${expr}` would not be legal JS syntax either, would it? That's an identifier followed by an object literal, which is a syntax error, I think.

**Comment by rzvxa** [comment]
> Actually, `${expr}` would not be legal JS syntax either, would it? That's an identifier followed by an object literal, which is a syntax error, I think.

It can be valid in the context of class/enum declaration:

```js
class ${ /* stuff */ }
```

**Comment by overlookmotel** [comment]
> ```js
> class ${ /* stuff */ }
> ```

Ha! Yes, of course you're right.

**Comment by hyf0** [comment]
https://x.com/_kermanx_/status/1817597119152439476

**Comment by 7086cmd** [comment]
Currently, I think we can utilize macros (`macro_rules!-like`) to parse the code for babel helper-like scenarios. The shared allocator can be utilized in parsing, so I do not anticipate a significant trade-off, particularly when the target code appears to be multiple lines, for instance, some polyfill-like code in CommonJS, async/await, etc, **for the time being**.

**Comment by overlookmotel** [comment]
Having thought about it more, the problem with an interpolation syntax is that we need to add support for it to the parser.

If we stick with valid (but unusual) syntax, we can avoid that problem. e.g.:

```rs
let expr = get_expression_somehow();
let stmt = ast!{ Statement: const foo = $expr; };
```

The remaining problem is how to deal with macro hygiene. `$expr` needs to be replaced in output with `expr`, where `expr`  resolves to `expr` outside the macro. Could look at `quote!` macro to see how it achieves this.

And then how to do the parsing at build-time (like`oxc_ast_tools`), instead of the compile-time impact of parsing in a proc macro.

**Comment by Dunqing** [comment]
Another question, how to set a correct Span?
